test/common/TestCommon.hs

```diff
@@ -9,15 +10,82 @@ import Prelude
 
 import qualified Data.Array.DynamicS as OT
 import qualified Data.Array.ShapedS as OS
+import           Data.IORef
 import qualified Data.Vector.Generic as V
 import           Numeric.LinearAlgebra (Vector)
 import qualified Numeric.LinearAlgebra as HM
+import           System.IO.Unsafe
 import           Test.Tasty
+import           Test.Tasty.HUnit
 import           Test.Tasty.QuickCheck
 
 import HordeAd hiding (sumElementsVectorOfDual)
 import HordeAd.Core.DualClass (Dual)
 
+-- Default value for eqEpsilonRef
+eqEpsilonDefault :: Double
+eqEpsilonDefault = 1e-6
+
+-- Ugly global epsilon used to compare floating point values.
+eqEpsilonRef :: IORef Double
+{-# NOINLINE eqEpsilonRef #-}
+eqEpsilonRef = unsafePerformIO $ newIORef eqEpsilonDefault
```

Comment:
I'm proud of you.

Reply:
It took me hours. It's so tricky! Did you know, for example, that this version:

```haskell
epsilonRef :: Fractional a => IORef a
{-# NOINLINE epsilonRef #-}
epsilonRef = unsafePerformIO $ newIORef 3.1415
```

doesn't work?! Even if I explicitly write `:: Double` everywhere I use it. Why is that?!

Reply:
I think you try to construct a value of an existential type `Exist a. Fractional a => IORef a`. However, to construct a value of an existential type you need to provide that type, not only a value. And you don't do that.

I'm unclear about the details of existential vs universal types in Haskell, e.g., if it's about positive occurrences of a type variable, or strictly positive and what happens if a variable has mixed variance. However, in this example it's simple. The occurrence of `a` is not under any number of arrows, so the whole thing is existential, not universal. If it was existential, it could be abused: you could construct a reference by initializing it with a Float, but later update it with a Double, overwriting a nearby memory cell (assuming Floats are smaller than Doubles).

Reply:
GHC implements type classes as dictionary passing, so

```haskell
epsilonRef :: Fractional a => IORef a
epsilonRef = unsafePerformIO $ newIORef 3.1415
```

is rather similar to

```haskell
epsilonRef :: Fractional a -> IORef a
epsilonRef fractionalDict = unsafePerformIO $ newIORef (fromRational fractionalDict 3.1415)
```

and it seems rather likely that this makes a fresh `IORef` containing `3.1415` at each use site. Would that explain what you see?

Reply:
> `Fractional a -> IORef a`

In that case, type variable `a` has an occurrence on the left had side of an arrow so, yes, this could probably be a universal type if interpreted that way. Similarly, if this function took an argument to initialize the ref with: `a -> IORef a`. And if these are lambdas, the body of a lambda is never evaluated, so the ref is not created. But I guess

```
epsilonRef :: IORef a
epsilonRef = unsafePerformIO $ newIORef undefined
```

would be firmly in the existential territory and no lambda, so perhaps my explanation only applied to that one. Not tested.

Reply:
@tomjaguarpaw I can't comment on GHC's type classes implementation but yes, what I see is like a fresh `IORef` each time.

Reply:
Here's a complete example:

```haskell
import Data.IORef
import System.IO.Unsafe

epsilonRef :: Fractional a => IORef a
{-# NOINLINE epsilonRef #-}
epsilonRef = unsafePerformIO $ newIORef 3.1415

setEpsilon :: Fractional a => a -> IO ()
setEpsilon = atomicWriteIORef epsilonRef

main :: IO ()
main = do
  () <- (setEpsilon :: (Double -> IO ())) (0.2 :: Double)  -- doesn't work!
  epsilon <- (readIORef (epsilonRef :: IORef Double))
  print $ "epsilon: " ++ show epsilon
```

On my machine this prints:

```
"epsilon: 3.1415"
```

Is this what you'd expect?